iml-agent/src/daemon_plugins/devices/devices.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    agent_error::ImlAgentError,
+    daemon_plugins::{devices::flat_devices::process_tree, DaemonPlugin, Output},
+};
+use futures::{
+    channel::oneshot, future, lock::Mutex, Future, FutureExt, Stream, StreamExt, TryFutureExt,
+    TryStreamExt,
+};
+use std::{collections::BTreeMap, pin::Pin, sync::Arc};
+use stream_cancel::{StreamExt as _, Trigger, Tripwire};
+use tokio::{io::AsyncWriteExt, net::UnixStream};
+use tokio_util::codec::{FramedRead, LinesCodec};
+
+fn device_stream() -> impl Stream<Item = Result<String, ImlAgentError>> {
+    UnixStream::connect("/var/run/device-scanner.sock")
+        .err_into()
+        .and_then(|mut conn| {
+            async {
+                conn.write_all(b"\"Stream\"\n")
+                    .err_into::<ImlAgentError>()
+                    .await?;
+
+                Ok(conn)
+            }
+        })
+        .map_ok(|c| FramedRead::new(c, LinesCodec::new()).err_into())
+        .try_flatten_stream()
+}
+
+pub fn create() -> impl DaemonPlugin {
+    Devices {
+        trigger: None,
+        state: Arc::new(Mutex::new((None, None))),
+    }
+}
+
+#[derive(Debug)]
+pub struct Devices {
+    trigger: Option<Trigger>,
+    state: Arc<Mutex<(Output, Output)>>,
+}
+
+impl DaemonPlugin for Devices {
+    fn start_session(
+        &mut self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        let (trigger, tripwire) = Tripwire::new();
+        let (tx, rx) = oneshot::channel();
+
+        self.trigger = Some(trigger);
+
+        let state = Arc::clone(&self.state);
+
+        tokio::spawn(
+            device_stream()
+                .boxed()
+                .and_then(|x| future::ready(serde_json::from_str(&x).map_err(|e| e.into())))
+                .and_then(|x| {
+                    let mut flat_devices = BTreeMap::new();
+
+                    process_tree(&x, None, &mut flat_devices);
+
+                    future::ready(
+                        serde_json::to_value(flat_devices)
+                            .map(Some)
+                            .map_err(|e| e.into()),
+                    )
+                })
+                .into_future()
+                .then(|(x, s)| {
```

Comment:
This is probably one of the most complicated parts of the PR.

Can't find what's `s` here.

Reply:
`s` here is the stream being consumed. 

Take a look at the [`into_future`](https://docs.rs/futures/0.3.1/futures/stream/trait.StreamExt.html#method.into_future) combinator which returns the next item in the stream and the stream itself.

The point of this is to avoid sending any data to the manager until the device-scanner itself has sent data.

We also only want a single device-scanner socket connection, instead of spinning up a new connection each tick of the plugin, so we need to make sure the stream itself get's spawned into a separate task.